File: android/src/main/kotlin/com/hodoan/barcode_scan_custom/CameraView.kt
package com.hodoan.barcode_scan_custom

import android.annotation.SuppressLint
import android.content.Context
import android.os.Build
import android.view.View
import androidx.annotation.RequiresApi
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.platform.PlatformView

@SuppressLint("RestrictedApi")
@RequiresApi(Build.VERSION_CODES.LOLLIPOP)
internal class CameraView(
    binaryMessenger: BinaryMessenger,
    context: Context,
    id: Int,
    creationParams: Any?
) :
    PlatformView, MethodChannel.MethodCallHandler, BarCodeCallback {
    private val cView: CView
    private val channelFlutter: MethodChannel
    private val channelNative: MethodChannel

    override fun getView(): View = cView

    override fun dispose() = cView.closeCamera()

    override fun onFlutterViewAttached(flutterView: View) {
        cView.openCamera()
        super.onFlutterViewAttached(flutterView)
    }

    init {
        channelFlutter = MethodChannel(binaryMessenger, "camera_view/flutter_$id")
        channelNative = MethodChannel(binaryMessenger, "camera_view/native_$id")
        channelNative.setMethodCallHandler(this)

        val c = Protos.FlutterConfiguration.parseFrom(creationParams as ByteArray)
//
//        val formats: List<Int> = creationParams?.get("formats") as List<Int>? ?: listOf()

        cView = CView(context, this, c)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "flash" -> result.success(cView.flash(call.arguments as Boolean))
            else -> result.notImplemented()
        }
    }

    override fun barcodeResult(barcode: Protos.ScanResult) {
        channelFlutter.invokeMethod("barcode", barcode.toByteArray())
    }
}

File: ios/Classes/CameraView.swift
//
//  CameraView.swift
//  barcode_scan_custom
//
//  Created by ominext on 13/03/2024.
//

import Foundation
import Flutter

class CameraView:NSObject,FlutterPlatformView{
    
    private var _view: CustomView
    
    private var channelFlutter: FlutterMethodChannel!
    private var channelNative: FlutterMethodChannel!
    
    init(
        frame: CGRect,
        viewIdentifier viewId: Int64,
        arguments args: Any?,
        binaryMessenger messenger: FlutterBinaryMessenger
    ) {
        let config = try? FlutterConfiguration(serializedData: args.parserData())
        
        _view = CustomView(frame: frame, config)
        
        channelFlutter = FlutterMethodChannel(name: "camera_view/flutter_\(viewId)", binaryMessenger: messenger)
        channelNative = FlutterMethodChannel(name: "camera_view/native_\(viewId)", binaryMessenger: messenger)
        
        super.init()
        
        _view.delegate = self
        channelNative.setMethodCallHandler(handle)
    }
    
    func view() -> UIView {
        return _view
    }
    
    func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "flash":
            _view.flash(call.arguments as! Bool)
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
}

extension CameraView: CustomViewDelegate{
    func barcode(_ result: ScanResult) {
        let data = try? result.serializedData()
        channelFlutter.invokeMethod("barcode", arguments: data)
    }
}
